src/extension/wps-download/src/main/java/org/geoserver/wps/gs/download/RasterDownload.java

```diff
@@ -207,188 +214,246 @@ public Resource execute(
                     reader.getFormat().getReadParameters();
             final List<GeneralParameterDescriptor> parameterDescriptors =
                     readParametersDescriptor.getDescriptor().descriptors();
-            Map<String, Serializable> coverageParameters = coverageInfo.getParameters();
-            GeneralParameterValue[] readParameters =
-                    CoverageUtils.getParameters(readParametersDescriptor, coverageParameters, false);
-
-            // read GridGeometry preparation and scaling setup if needed
-            GridGeometry2D requestedGridGeometry = null;
-            boolean isImposedTargetSize = targetSizeX != null || targetSizeY != null;
-
-            if (targetSizeX == null && targetSizeY == null) {
-                // No size is specified. Just do a read and reproject (if needed) + a final crop
-                if (LOGGER.isLoggable(Level.FINE)) {
-                    LOGGER.log(
-                            Level.FINE,
-                            "No Target size has been specified. Requested GridGeometry "
-                                    + "will be automatically computed");
-                }
-                GridGeometryProvider provider = new GridGeometryProvider(crsRequestHandler);
-                requestedGridGeometry = provider.getGridGeometry();
-
-                if (LOGGER.isLoggable(Level.FINE)) {
-                    LOGGER.log(Level.FINE, "Computed requested GridGeometry: " + requestedGridGeometry.toString());
-                }
-                readParameters = CoverageUtils.mergeParameter(
-                        parameterDescriptors,
-                        readParameters,
-                        requestedGridGeometry,
-                        AbstractGridFormat.READ_GRIDGEOMETRY2D.getName().getCode());
+            Integer[] targetSize = {targetSizeX, targetSizeY};
+
+            RasterDownloadContext ctx = new RasterDownloadContext(coverageInfo, targetSize, crsRequestHandler);
+            ctx.computeGridGeometry(reader, crsRequestHandler);
+
+            // Set optional context params
+            ctx.bandIndices = bandIndices;
+            ctx.interpolation = interpolation;
+            ctx.roi = roi;
+            ctx.clip = clip;
+            ctx.filter = filter;
+            ctx.targetVerticalCRS = targetVerticalCRS;
+
+            GeneralParameterValue[] readParameters = initializeReadParameters(
+                    reader,
+                    coverageInfo,
+                    crsRequestHandler,
+                    useTargetCrsAsNative,
+                    ctx.requestedGridGeometry,
+                    parameterDescriptors);
+
+            ctx.computeBackgroundValues(readParameters);
+            slicer = SpatioTemporalCoverageSlicer.SpatioTemporalCoverageSlicerHolder.get();
+            if (slicer == null) {
+                slicer = new SpatioTemporalCoverageSlicer(reader, coverageInfo, filter);
+                slicer.initialize();
+            }
```

Comment:
I get the intention (avoid querying the same granule source twice) but it has a potential to go sideways. Clients like mapstore call the estimator directly before calling the download process... is this thread local cleared also in that case?

Generally speaking, we need some mechanism to ensure that the  thread local gets removed no matter what, regardless of whether the process runs in foreground or in an async thread, like org.geoserver.wps.ProcessListener, registered as a bean in the application context.

Also, I'd recommend moving all the setup machinery in the part of code asking to grab the slicer, to aviod duplications.

Reply:
Added a SlicerCleanupListener that takes care of removing the ThreadLocal.
Initialization also moved in the slicer creation.